server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java

```diff
@@ -71,7 +74,9 @@ public String getBase64UUID()  {
                 timestamp++;
             }
 
-            lastTimestamp = timestamp;
+            if (this.lastTimestamp.compareAndSet(lastTimestamp, timestamp)) {
```

Comment:
I think we can fold the loop into a convenience method on `AtomicLong`:

```diff
diff --git a/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java b/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
index c30a8d0aaa2..aa14d608aa7 100644
--- a/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+++ b/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
@@ -21,6 +21,7 @@ package org.elasticsearch.common;
 
 import java.util.Base64;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * These are essentially flake ids but we use 6 (not 8) bytes for timestamp, and use 3 (not 2) bytes for sequence number. We also reorder
@@ -37,7 +38,7 @@ class TimeBasedUUIDGenerator implements UUIDGenerator {
     private final AtomicInteger sequenceNumber = new AtomicInteger(SecureRandomHolder.INSTANCE.nextInt());
 
     // Used to ensure clock moves forward:
-    private long lastTimestamp;
+    private AtomicLong lastTimestamp = new AtomicLong();
 
     private static final byte[] SECURE_MUNGED_ADDRESS = MacAddressProvider.getSecureMungedAddress();
 
@@ -58,21 +59,22 @@ class TimeBasedUUIDGenerator implements UUIDGenerator {
     @Override
     public String getBase64UUID()  {
         final int sequenceId = sequenceNumber.incrementAndGet() & 0xffffff;
-        long timestamp = currentTimeMillis();
 
-        synchronized (this) {
-            // Don't let timestamp go backwards, at least "on our watch" (while this JVM is running).  We are still vulnerable if we are
-            // shut down, clock goes backwards, and we restart... for this we randomize the sequenceNumber on init to decrease chance of
-            // collision:
-            timestamp = Math.max(lastTimestamp, timestamp);
+        final long timestamp = lastTimestamp.updateAndGet(
+            last -> {
+                // Don't let timestamp go backwards, at least "on our watch" (while this JVM is running).  We are still vulnerable if we are
+                // shut down, clock goes backwards, and we restart... for this we randomize the sequenceNumber on init to decrease chance of
+                // collision:
+                long nonBackwardsCurrentTimestamp = Math.max(last, currentTimeMillis());
 
-            if (sequenceId == 0) {
-                // Always force the clock to increment whenever sequence number is 0, in case we have a long time-slip backwards:
-                timestamp++;
-            }
+                if (sequenceId == 0) {
+                    // Always force the clock to increment whenever sequence number is 0, in case we have a long time-slip backwards:
+                    nonBackwardsCurrentTimestamp++;
+                }
 
-            lastTimestamp = timestamp;
-        }
+                return nonBackwardsCurrentTimestamp;
+            }
+         );
 
         final byte[] uuidBytes = new byte[15];
         int i = 0;
```

Reply:
In your proposal we might call currentTimeMillis() multiple times, but I don't think we have to? It could be computed only once like in master or Tim's proposal?

Reply:
I missed that, thanks. It can lifted out of the `updateAndGet` update function as @tbrooks8 did in bf13356b18bcb22e0da0db8a861afae00cd5847b.